x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,728 @@
+
+- **Update Rule** - `PUT /rules`: same as Patch Rule.
+    - Used **only** when updating/modifying a single rule via the Rule Details page
+- **Patch Rule** - `PATCH /rules`: should migrate patched rule to prevent the endpoint from returning an outdated (non-migrated) rule.
+    - Used for attaching shared exceptions list to rules
+- **Bulk Update Rules** - `PUT /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Bulk Patch Rules** - `PATCH /rules/_bulk_update`: same as for Patch Rule
+    - Deprecated and unused by the UI (might still be used by API users)
+- **Import Rules** - `POST /rules/_import`: 
+    - if the user imports rules with the updated rule schema (`prebuilt` and `customized` fields are present), then save rule as-is. 
+    - if the user imports a "legacy" rule (with the `immutable` field, and no `prebuilt` or `customized` fields), the `immutable` field will be saved for backwards compatibility as well, and the `prebuilt` and `customized` fields should always be set to `false` (importing a non-migrated rule means that the rule could not be modified or exported in the Kibana version where it was exported from, so it could not have been a prebuilt rule).
+- **Bulk Actions** - `POST /rules/_bulk_action`: 
+    - This endpoint includes many types of actions, and only a subset of them should perform a migration.
+    - This endpoint also includes a `dry_run` mode that is executed to evaluate preconditions and warn the user before executing the actual request. No migration logic should take place for dry run requests, i.e when `dry_run=true`.
+    - Actions that should perform migration logic:
+        - Edit:
+            - endpoint should migrate rule while performing migration. We can take advantage of the `ruleParamsModifier` to carry out the migration, regardless of the type of edit that is being performed.
+    - Actions that should **not** perform migration logic:
+        - Enable
+        - Disable
+        - Delete
+        - Export:
+            - export endpoint logic does not write/modify rules, only reads them and makes them available for download. Adding new writing logic could have a big impact on its performance.
+            - we will now allow the export of prebuilt rules
+            - export endpoint should therefore only export the rule in their current state (migrated or non-migrated), and the import endpoint will take care of the migration logic
+        - Duplicate:
+            - this action should not migrate the original rule being duplicated, as it would violate the principle of least surprise
+            - the new rule being created should have the new schema
+            - we should only update the duplication logic to account for possibility that the new `prebuilt` and `customized` fields already exist in the original rule
```

Comment:
Nit: We will need to handle this case in pretty much every endpoint.

Reply:
Yes, removing this from here, it will be clarified with code in the implementation details of this endpoint.